### PR TITLE
Adding match_pattern param to changelog_from_git_commits

### PR DESCRIPTION
**Pull request by 7mllm7**
Adds the option to fetch changelog from git commits with _from_ tag name that matches a given [glob(7)](http://man7.org/linux/man-pages/man7/glob.7.html) pattern.
For example if you're tags have a naming convention like:
- `builds/production/v1.6` - latest for **production** environment, and latest in repo/branch
- `builds/dev/v1.5` - latest for **dev** environment, but _not_ latest in repo/branch

So, if you call `changelog_from_git_commits(match_pattern: 'builds/dev/*')`, then it will skip `builds/production/v1.6` although it is the latest, and take `builds/dev/v1.5` for _from_ tag name.

### CODE REVIEW

**Comment by 7mllm7** [comment]
Anyone has an idea why this would hang-fail only on Ruby 2.1.6?

**Comment by micnguyen** [comment]
That's actually a great addition that I can see much utility for me in the future. 

I'm not sure why it's failing on 2.1.6 - the build log say because you're running with 2.1.7 syntax but looking at your commit (and not being a Ruby guy myself) I can't tell where it might be without a more verbose logs.

**Comment by mfurtak** [comment]
Hi @7mllm7 - thanks for submitting this! :+1: 

I think this is a great idea and I'd like to get this kind of support in there. #1031 is also out there, and I think the goals are similar between them.

Let me confirm with @scottrhoyt about whether `changelog_from_git_commits` can meet his needs in this regard, and then I think we can put something together that solves it all! :100:

File: docs/Actions.md [hunk 1]
```diff
@@ -1126,6 +1126,7 @@ changelog_from_git_commits
 changelog_from_git_commits(
   between: ['7b092b3', 'HEAD'], # Optional, lets you specify a revision/tag range between which to collect commit info
   pretty: '- (%ae) %s', # Optional, lets you provide a custom format to apply to each commit when generating the changelog text
+  match_pattern: nil, # Optional, lets you search for a tag name that matches a glov(7) pattern
```
Comment (mfurtak) [inline docs/Actions.md]: typo: glov -> glob

File: lib/fastlane/actions/changelog_from_git_commits.rb [hunk 2]
```diff
@@ -56,6 +56,12 @@ def self.available_options
                                        optional: true,
                                        default_value: '%B',
                                        is_string: true),
+          FastlaneCore::ConfigItem.new(key: :match_pattern,
+                                       env_name: 'FL_CHANGELOG_FROM_GIT_COMMITS_MATCH_PATTERN',
+                                       description: 'Whether or not to match a glob(7) pattern in tag name',
```
Comment (mfurtak) [inline lib/fastlane/actions/changelog_from_git_commits.rb]: Can we rephrase this as: 'A glob(7) pattern to match against when finding the last git tag'

File: lib/fastlane/actions/changelog_from_git_commits.rb [hunk 3]
```diff
@@ -56,6 +56,12 @@ def self.available_options
                                        optional: true,
                                        default_value: '%B',
                                        is_string: true),
+          FastlaneCore::ConfigItem.new(key: :match_pattern,
```
Comment (mfurtak) [inline lib/fastlane/actions/changelog_from_git_commits.rb]: Can we call this param `:tag_match_pattern` so that it is a little more clear what we're matching against?

File: lib/fastlane/actions/changelog_from_git_commits.rb [hunk 4]
```diff
@@ -56,6 +56,12 @@ def self.available_options
...
+                                       env_name: 'FL_CHANGELOG_FROM_GIT_COMMITS_MATCH_PATTERN',
+                                       description: 'Whether or not to match a glob(7) pattern in tag name',
+                                       optional: true,
+                                       default_value: nil,
```
Comment (mfurtak) [inline lib/fastlane/actions/changelog_from_git_commits.rb]: `nil` will be the default `default_value`, so this can be omitted. :+1:

File: lib/fastlane/actions/changelog_from_git_commits.rb [hunk 5]
```diff
@@ -56,6 +56,12 @@ def self.available_options
...
+                                       is_string: false),
```
Comment (mfurtak) [inline lib/fastlane/actions/changelog_from_git_commits.rb]: This option is a `String`, yes? If so, `is_string` will default to `true` so this can be omitted.

File: lib/fastlane/helper/git_helper.rb [hunk 6]
```diff
@@ -6,10 +6,11 @@ def self.git_log_between(pretty_format, from, to)
       nil
     end
 
-    def self.last_git_tag_name(match_lightweight = true)
+    def self.last_git_tag_name(match_lightweight = true, match_pattern = nil)
       command = ['git describe']
       command << '--tags' if match_lightweight
       command << '--abbrev=0'
+      command << "--match #{match_pattern}" if match_pattern
```
Comment (mfurtak) [inline lib/fastlane/helper/git_helper.rb]: I think you may need `${match_pattern.shellescape}` here, though it makes the value you need to test for in the `changelog_from_git_commits_spec` wind up double-escaped and fairly silly looking :smile:

**Comment by mfurtak** [comment]
Hi @7mllm7 - Thanks again for putting this together, and sorry for the sudden flurry of comments!

The CI hang is something that happens to us every now and again, and I don't think it's related to your changes. It seems happy enough after restarting the job, though now there's the matter of merge conflicts. :sweat_smile: 

It would be great to get this in, so please let me know if anything I've said needs clarification. :+1:

**Comment by 7mllm7** [comment]
Ohhh, I implemented the code review, but I'm used to squash-rebase, delete and then re-push the feature branch, so it closed the pull request.

__Sorry__ about that! so what can I do? create another pull request? reopen this one somehow?

Thanks!

**Comment by mfurtak** [comment]
@7mllm7 Oh, no problem. Opening a new PR is certainly fine!

Just as a heads up, we're continuing to have discussions here about how to reconcile all the recent improvements we've been seeing around this action. There's #1031, which I think is very compatible with this in the long term. However, there's also #1040 which got in first. I believe that one conflicts a bit with the use of the `--match` flag.

So, in summary, we still want to get this in -- it just make take a little time for us to revisit it. Thank you! :+1:

**Comment by 7mllm7** [comment]
Right, thanks for the cr and clarification, I've opened https://github.com/fastlane/fastlane/pull/1105.

Next time, would a merge be the better technique?

**Comment by 7mllm7** [comment]
Regarding https://github.com/fastlane/fastlane/pull/1040, I updated mine to also support it.
